orangecontrib/text/widgets/owconcordance.py

```diff
@@ -0,0 +1,218 @@
+from orangecontrib.text.corpus import Corpus
+from orangecontrib.text.preprocess import Preprocessor, WordPunctTokenizer
+
+
+class IO:
+    DATA = 'Corpus'
+
+
+class ConcordanceModel(QAbstractTableModel):
+    "A model for constructing concordances from text."
+    def __init__(self):
+        QAbstractTableModel.__init__(self)
+        self.word = None
+        self.data = None
+        self.indices = None
+        self.doc_count = None
+
+    def set_word(self, word):
+        self.modelAboutToBeReset.emit()
+        self.word = word
+        self._compute_indices()
+        self.modelReset.emit()
+
+    def set_data(self, data):
+        self.modelAboutToBeReset.emit()
+        self.data = data
+        self._compute_indices()
+        self.modelReset.emit()
+
+    def flags(self, _):
+        return Qt.ItemIsEnabled | Qt.ItemIsSelectable
+
+    def rowCount(self, parent=None, *args, **kwargs):
+        return len(self.doc_count) if self.doc_count is not None else 0
+        # return len(self.indices) if self.indices is not None else 0
+
+    def columnCount(self, parent=None, *args, **kwargs):
+        return 3
+
+    def data(self, index, role=Qt.DisplayRole):
+        row, col = index.row(), index.column()
+        width = 5
+
+        if role == Qt.DisplayRole:
+            doc = row
+            beg = ''
+            mid = ''
+            end = ''
+            for i_doc, index in self.indices:
+                if i_doc == doc:
+                    beg += ' '.join(self.data.tokens[doc][index-width:index]) + '\n'
+                    mid += self.data.tokens[doc][index] + '\n'
+                    end += ' '.join(self.data.tokens[doc][index+1:index+width]) + '\n'
+
+            if col == 0:
+                return beg
+            if col == 1:
+                return mid
+            if col == 2:
+                return end
+
+            # doc, index = self.indices[row]
+            # if col == 0:
+            #     return ' '.join(self.tokens[doc][index-width:index])
+            # if col == 1:
+            #     return self.tokens[doc][index]
+            # if col == 2:
+            #     return ' '.join(self.tokens[doc][index+1:index+width])
+
+        elif role == Qt.TextAlignmentRole:
+            return [Qt.AlignRight | Qt.AlignVCenter, Qt.AlignCenter, Qt.AlignLeft | Qt.AlignVCenter][col]
+
+        elif role == Qt.ForegroundRole:
+            if col == 1:
+                return QBrush(QColor(0, 0, 255))
+
+    def _compute_indices(self):  # type: () -> Optional[None, list]
+        if self.data is None or self.word is None:
+            self.indices = None
+            return
+        if self.data and not self.data.has_tokens():
+            preprocessor = Preprocessor(tokenizer=WordPunctTokenizer())
+            preprocessor(self.data)
+        self.indices = [(i, j) for i, doc in enumerate(self.data.tokens)
+                        for j, w in enumerate(doc) if w.lower() == self.word.lower()]
```

Comment:
AFAIK it builds an index. Once build, getting the indices of words occurrences in a document should take constant time rather than linear in regard to the number of tokens in the document.

It does not take a word on input — it builds an index for all words — so I guess it could be a part of `set_data`.

Yup, the tokenization should definitely be a part of `set_data`. Though, am still not sure, if it wouldn't be better to disable lowercasing in the default preprocessor. Then we wouldn't need to bother about preprocessing at all but could just call `corpus.tokens`.